File: dotnet/Aoc/Aoc.Spaceship/Computer/IntcodeComputer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Aoc.Spaceship.Computer.Instructions;

namespace Aoc.Spaceship.Computer
{
    public class IntcodeComputer : IInputSource
    {
        private int[]? _memory;
        private List<int> _input;
        private int _inputCounter;
        private int _instructionPointer;
        //Keeps track of how many times we've asked parent computer for input
        private int _outputCounter = 0;
        public IInputSource InputPort { get; set; }
        public string Name { get; set; }
        public List<int>? Output { get; set; }

        public IntcodeComputer() : this(new int[] {})
        {
        }

        public IntcodeComputer(int[] input)
        {
            _input = input.ToList();
        }

        public int[] RunProgram(int[] program)
        {
            return RunProgram(program, new int[]{});
        }

        public int[] RunProgram(int[] program, int input)
        {
            return RunProgram(program, new [] {input});
        }

        public int[] RunProgram(int[] program, int[] input)
        {
            return RunProgramAsync(program, input).Result;
        }

        public async Task<int[]> RunProgramAsync(int[] program, int[] input)
        {
            Initialize(program, input.ToList());

            //Main computer logic
            _instructionPointer = 0;
            while (_instructionPointer < _memory.Length)
            {
                var instruction = GetNextInstruction();
                if (instruction is Halt)
                    return _memory;
                _instructionPointer = await ExecuteInstructionAsync(instruction);
            }

            //We should always see a halt operation at the end of the program
            throw new InvalidIntcodeProgram("No halt instruction at end of program");
        }

        private Instruction GetNextInstruction()
        {
            var rawOpcode = _memory[_instructionPointer];
            Opcodes opcode = (Opcodes) (rawOpcode % 100);
            Instruction instruction;
            switch (opcode)
            {
                case Opcodes.Halt:
                    instruction = new Halt();
                    break;
                case Opcodes.Display:
                    instruction = new Display();
                    break;
                case Opcodes.Put:
                    instruction = new Put(rawOpcode);
                    break;
                case Opcodes.JumpIfTrue:
                    instruction = new Jump(rawOpcode,() => true);
                    break;
                case Opcodes.JumpIfFalse:
                    instruction = new Jump(rawOpcode,() => false);
                    break;
                case Opcodes.LessThan:
                    instruction = new Compare(rawOpcode, compareFunction: (x, y) => x < y);
                    break;
                case Opcodes.Equals:
                    instruction = new Compare(rawOpcode, compareFunction: (x, y) => x == y);
                    break;
                case Opcodes.Add:
                    instruction = new Instructions.Math(rawOpcode, mathOperation: (x, y) => x + y);
                    break;
                case Opcodes.Multiply:
                    instruction = new Instructions.Math(rawOpcode, mathOperation: (x, y) => x * y);
                    break;
                default:
                    throw new InvalidIntcodeProgram($"Opcode {opcode} unknown");
            }
            if (_memory.Length < _instructionPointer + instruction.Length)
                throw new InvalidIntcodeProgram("Last instruction is incomplete");
            return instruction;
        }

        private async Task<int> ExecuteInstructionAsync(Instruction instruction)
        {
            switch (instruction)
            {
                case Instructions.Math mathInstruction:
                    ExecuteInstruction(mathInstruction);
                    break;
                case Display displayInstruction:
                    ExecuteInstruction(displayInstruction);
                    break;
                case Put putInstruction:
                    await ExecuteInstructionAsync(putInstruction);
                    break;
                case Jump jumpInstruction:
                    return ExecuteInstruction(jumpInstruction);
                case Compare compareInstruction:
                    ExecuteInstruction(compareInstruction);
                    break;
                default: 
                    throw new InvalidIntcodeProgram($"Unknown instruction {instruction}");
            }
            return _instructionPointer + instruction.Length;
        }

        private int ExecuteInstruction(Jump instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            if (instruction.ShouldJump(parameter1))
                return parameter2;
            return _instructionPointer + instruction.Length;
        }

        private void ExecuteInstruction(Compare instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            var destinationAddress = GetParameterValue(instruction, 3);
            _memory[destinationAddress] = instruction.CompareFunction(parameter1, parameter2) ? 1 : 0;
        }

        private void ExecuteInstruction(Instructions.Math instruction)
        {
            int parameter1 = GetParameterValue(instruction, 1);
            int parameter2 = GetParameterValue(instruction, 2);
            var instructionValue = instruction.MathOperation(parameter1, parameter2);
            var destinationAddress = GetParameterValue(instruction, 3);
            _memory[destinationAddress] = instructionValue;
        }

        private void ExecuteInstruction(Display instruction)
        {
            var outputValue = GetParameterValue(instruction, 1);
            Output.Add(outputValue);
        }

        private async Task ExecuteInstructionAsync(Put instruction)
        {
            var input = await GetInput();
            if (instruction.ParameterModes[0] == ParameterMode.Immediate)
                _memory[_instructionPointer + 1] = input;
            else
                _memory[_memory[_instructionPointer + 1]] = input;
        }

        //Allows the computer to act as input for another computer
        public async Task<int> GetInput(int outputCounter)
        {
            while (Output == null || Output.Count - 1 < outputCounter)
            {
                await Task.Delay(1);
            }
            return Output[outputCounter];
        }

        private async Task<int> GetInput()
        {
            if (_inputCounter > _input.Count - 1 && !(InputPort is null))
                return await GetFromInputPort();
            return GetFromInitialInput();
        }

        private async Task<int> GetFromInputPort()
        {
            //Getting input from whatever we have plugged in to the input port
            var inputValue = await InputPort.GetInput(_outputCounter);
            _outputCounter++;
            return inputValue;
        }

        private int GetFromInitialInput()
        {
            //This is input that was given at the beginning of the run
            var inputValue = _input[_inputCounter];
            _inputCounter++;
            return inputValue;
        }

        private int GetParameterValue(Instruction instruction, int parameterPosition)
        {
            if (instruction.ParameterModes[parameterPosition - 1] == ParameterMode.Immediate)
                return _memory[_instructionPointer + parameterPosition];
            return _memory[_memory[_instructionPointer + parameterPosition]];
        }

        private void Initialize(int[] program, List<int> input)
        {
            //Make sure we don't use input from previous runs
            _input = input;
            _inputCounter = 0;
            //Make sure we don't have output from previous program runs
            Output = new List<int>();
            //Don't surprise the user and make changes to the incoming program
            _memory = Array.Empty<int>();
            Array.Resize(ref _memory, program.Length);
            program.CopyTo(_memory, 0);
        }
    }
}

File: python/aoc/spaceship/computer/intcode_computer.py
from spaceship.computer.instructions import *


class IntcodeComputer(object):
    def __init__(self):
        self._instruction_pointer: int = 0
        self._memory: list[int]
        self._input: int
        self._Output: list[int]

    @property
    def output(self):
        return self._Output

    def run_program(self, program: list[int], user_input) -> list[int]:
        self.__initialize(program, user_input)

        # Main computer logic
        while self._instruction_pointer < len(program):
            instruction = self.__get_next_instruction()
            if isinstance(instruction, Halt):
                return self._memory
            self._instruction_pointer = self.__execute_instruction(instruction)

        raise InvalidIntcodeProgram("No halt instruction at end of program")

    def __get_next_instruction(self) -> Instruction:
        opcode = self._memory[self._instruction_pointer]
        opcode_enum = self.__get_opcode(opcode)
        match opcode_enum:
            case Opcodes.HALT:
                return Halt()
            case Opcodes.DISPLAY:
                return Display()
            case Opcodes.PUT:
                return Put(opcode)
            case Opcodes.JUMPIFTRUE:
                return Jump(opcode, lambda: True)
            case Opcodes.JUMPIFFALSE:
                return Jump(opcode, lambda: False)
            case Opcodes.LESSTHAN:
                return Compare(opcode, lambda x, y: x < y)
            case Opcodes.EQUALS:
                return Compare(opcode, lambda x, y: x == y)
            case Opcodes.ADD:
                return Math(opcode, lambda x, y: x + y)
            case Opcodes.MULTIPLY:
                return Math(opcode, lambda x, y: x * y)

    def __execute_instruction(self, instruction: Instruction) -> int:
        if isinstance(instruction, Math):
            self.__execute_math_instruction(instruction)
        if isinstance(instruction, Display):
            self.__execute_display_instruction(instruction)
        if isinstance(instruction, Put):
            self.__execute_put_instruction(instruction)
        if isinstance(instruction, Compare):
            self.__execute_compare_instruction(instruction)
        if isinstance(instruction, Jump):
            return self.__execute_jump_instruction(instruction)
        return self._instruction_pointer + instruction.length

    def __execute_jump_instruction(self, instruction: Jump) -> int:
        parameter1 = self.__get_parameter_value(instruction, 1)
        parameter2 = self.__get_parameter_value(instruction, 2)
        if instruction.should_jump(parameter1):
            return parameter2
        return self._instruction_pointer + instruction.length

    def __execute_compare_instruction(self, instruction: Compare) -> None:
        parameter1 = self.__get_parameter_value(instruction, 1)
        parameter2 = self.__get_parameter_value(instruction, 2)
        destination_address = self.__get_parameter_value(instruction, 3)
        self._memory[destination_address] = (1 if instruction.compare_function(parameter1, parameter2) else 0)

    def __execute_math_instruction(self, instruction: Math) -> None:
        parameter1 = self.__get_parameter_value(instruction, 1)
        parameter2 = self.__get_parameter_value(instruction, 2)
        instruction_value = instruction.math_operation(parameter1, parameter2)
        destination_address = self.__get_parameter_value(instruction, 3)
        self._memory[destination_address] = instruction_value

    def __execute_display_instruction(self, instruction: Display) -> None:
        self._Output.append(self.__get_parameter_value(instruction, 1))

    def __execute_put_instruction(self, instruction: Put) -> None:
        if self._input is None:
            raise InvalidIntcodeProgram("This program expects input from user and none was given")
        if instruction.parameter_modes[0] == ParameterMode.IMMEDIATE:
            self._memory[(self._instruction_pointer + 1)] = self._input
        else:
            self._memory[self._memory[(self._instruction_pointer + 1)]] = self._input

    def __get_parameter_value(self, instruction: Instruction, parameter_position: int) -> int:
        memory_location = self.__get_memory_location(instruction, parameter_position)
        if memory_location > len(self._memory) - 1:
            raise InvalidIntcodeProgram("The program tried to access memory that does not belong to it")
        return self._memory[memory_location]

    def __get_memory_location(self, instruction: Instruction, parameter_position: int) -> int:
        if instruction.parameter_modes[(parameter_position - 1)] == ParameterMode.IMMEDIATE:
            return self._instruction_pointer + parameter_position
        else:
            return self._memory[self._instruction_pointer + parameter_position]

    def __get_opcode(self, opcode: int) -> Opcodes:
        try:
            return Opcodes(opcode % 100)
        except Exception:
            raise InvalidIntcodeProgram("Unknown opcode {}".format(opcode))

    def __initialize(self, program, user_input) -> None:
        self._input = user_input
        self._instruction_pointer = 0
        # Don't surprise the user and change the incoming program
        self._Output = []
        self._memory = program.copy()


class InvalidIntcodeProgram(Exception):
    pass
